polar-core/src/vm.rs

```diff
@@ -2159,15 +2160,36 @@ impl PolarVirtualMachine {
     ///  - Failure => backtrack
     fn unify(&mut self, left: &Term, right: &Term) -> PolarResult<()> {
         match (left.value(), right.value()) {
-            (Value::Expression(_), _) | (_, Value::Expression(_)) => {
-                return Err(self.type_error(
-                    left,
-                    format!(
-                        "cannot unify expressions directly `{}` = `{}`",
-                        left.to_polar(),
-                        right.to_polar()
-                    ),
-                ));
+            (Value::Expression(op), other) | (other, Value::Expression(op)) => {
+                match op {
```

Comment:
That would exacerbate the performance issue even more though.

Reply:
Fair point on the rewriting all being in the VM. Let's leave it. FWIW this comment is coming from wanting to make the partial representation internal & not something the VM needs to worry about so that we are free to change to a more efficient representation than a vec of constraints. But, there are a bunch of other things that also have to happen to get there.

I'm not sure I follow why though on the performance issue. Because we'd iterate over the entire partial to produce the query goal?